main/src/main/java/com/kenshoo/pl/entity/internal/DbCommandsOutputGenerator.java

```
@@ -145,28 +154,31 @@ private CreateRecordCommand newCreateRecord(EntityChange<E> entityChange) {
 
     @Override
     public Stream<? extends EntityField<?, ?>> requiredFields(Collection<? extends EntityField<E, ?>> fieldsToUpdate, ChangeOperation changeOperation) {
-        // If update, find which secondary tables are affected.
-        // For those secondary tables take their foreign keys to primary, translate referenced fields of primary to EntityFields and add them to fields to fetch
-        if (changeOperation == ChangeOperation.UPDATE) {
-            DataTable primaryTable = entityType.getPrimaryTable();
-            if (isFieldsInSecondaryTables(fieldsToUpdate, primaryTable)) {
-                return getPrimaryKeyFields(entityType);
-            }
+        if (changeOperation != UPDATE || fieldsToUpdate.stream().noneMatch(this::isSecondaryField)) {
+            return Stream.empty();
         }
+        var secondaryTables = secondaryTables(fieldsToUpdate).collect(toList());
+        return Stream.concat(seq(secondaryTableAlreadyExistChecker.fieldsToFetch(secondaryTables)), keysFromPrimaryTableFor(secondaryTables));
+    }
 
-        return Stream.empty();
+    private Seq<? extends EntityField<E, ?>> keysFromPrimaryTableFor(List<DataTable> secondaryTables) {
+        return seq(secondaryTables).flatMap(table -> {
+            var foreignKey = table.getForeignKey(entityType.getPrimaryTable());
+            var primaryFields = foreignKey.getKey().getFields();
+            return seq(primaryFields)
+                    .map(field-> entityType.findField(field).get());
```

@dimagorelik , NOTE: before the fix, we always asked for the PK assuming the 2nry table always referred the primary using the PK. In this PR, we also fix this assumption by requiring the real fields.

This specific line will throw an exception if the 2nry table is using a field that is not defined by the EntityType.
Let's throw an explicit exception:
`.get()` --> `.orElseThow(() -> "field " + field +" is an FK from table " + secondary + " to " + primary + " but is not defined on entity type " + entityType);`